types/wlr_seat.c

```diff
@@ -0,0 +1,314 @@
+};
+
+static void seat_destroy(struct wl_resource *resource) {
+	struct wlr_seat *seat = wl_resource_get_user_data(resource);
+	for(unsigned int i = 0; i < seat->resources.length; ++i) {
+		struct wlr_seat_resources *resources = seat->resources.items[i];
+		if (resources->seat == resource) {
+			if (resources == seat->pointer_over_res) {
+				seat->pointer_over_res = NULL;
+			}
+
+			if (resources == seat->keyboard_focus_res) {
+				seat->keyboard_focus_res = NULL;
+			}
+
+			list_del(&seat->resources, i);
+			free(resources);
+			break;
+		}
+	}
+}
+
+static void seat_bind(struct wl_client *client, void *data, uint32_t version, uint32_t id) {
+	struct wlr_seat *seat = data;
+	assert(client && seat);
+	if (version > 6) {
+		wlr_log(L_ERROR, "Client requested unsupported wl_seat version, disconnecting");
+		wl_client_destroy(client);
+		return;
+	}
+
+	struct wl_resource *resource = wl_resource_create(
+			client, &wl_seat_interface, version, id);
+	if (!resource) {
+		wlr_log(L_ERROR, "Failed to create seat resource");
+		wl_client_post_no_memory(client);
+		return;
+	}
+
+	struct wlr_seat_resources *resources;
+	if (!(resources = calloc(1, sizeof(*resources)))) {
+		wlr_log(L_ERROR, "Failed to alloc seat resources");
+		wl_resource_destroy(resource);
+		wl_client_post_no_memory(client);
+		return;
+	}
+
+	wl_resource_set_implementation(resource, &seat_implementation, seat, seat_destroy);
+	wl_seat_send_capabilities(resource, seat->caps);
+	resources->seat = resource;
+	list_add(&seat->resources, resources);
+
+	if (version >= 2) {
+		wl_seat_send_name(resource, seat->name);
+	}
+}
+
+struct wlr_seat *wlr_seat_create(struct wl_display *display, const char *name) {
+	struct wlr_seat *seat;
+	if (!(seat = calloc(1, sizeof(struct wlr_seat)))) {
+		wlr_log(L_ERROR, "Failed to allocate seat");
+		return NULL;
+	}
+
+	if (name == NULL) {
+		const char *env_name = getenv("XDG_SEAT");
+		name = env_name ? env_name : "seat0";
+	}
+
+	seat->global = wl_global_create(display, &wl_seat_interface, 6, seat, &seat_bind);
+	seat->name = name;
+	return seat;
+}
+
+void wlr_seat_destroy(struct wlr_seat *seat) {
+	wl_global_destroy(seat->global);
+	free(seat);
+}
+
+void wlr_seat_set_caps(struct wlr_seat *seat, uint32_t caps) {
+	seat->caps = caps;
+	for(unsigned int i = 0; i < seat->resources.length; ++i) {
+		struct wlr_seat_resources *resources = seat->resources.items[i];
+		wl_seat_send_capabilities(resources->seat, caps);
+	}
+}
+
+void wlr_seat_pointer_move(struct wlr_seat *seat, uint32_t time, wl_fixed_t surface_x,
+		wl_fixed_t surface_y) {
+	if (!seat->pointer_over_res || !seat->pointer_over_res->pointer) {
+		return;
+	}
+
+	wl_pointer_send_motion(seat->pointer_over_res->pointer, time, surface_x, surface_y);
+}
+
+void wlr_seat_pointer_enter(struct wlr_seat *seat, uint32_t serial,
+		struct wl_resource *surface, wl_fixed_t surface_x, wl_fixed_t surface_y) {
+	// TODO: send pointer_leave event if pointer_over_* is not NULL?
```

Comment:
This interface should be dumb and should just execute the orders of the caller. It shouldn't try to manage focus *at all*.